src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -516,21 +535,67 @@ private static Expression GetValueFromProperty(Expression sourceExpression, stri
             return Expression.Convert(indexExpression, typeof(string));
         }
 
+        private static Expression BindParameterFromService(ParameterInfo parameter)
+        {
+            var nullability = nullabilityContext.Create(parameter);
+            var isOptional = parameter.HasDefaultValue || nullability.ReadState == NullabilityState.Nullable;
+
+            return isOptional
+                ? Expression.Call(GetServiceMethod.MakeGenericMethod(parameter.ParameterType), RequestServicesExpr)
+                : Expression.Call(GetRequiredServiceMethod.MakeGenericMethod(parameter.ParameterType), RequestServicesExpr);
+        }
+
         private static Expression BindParameterFromValue(ParameterInfo parameter, Expression valueExpression, FactoryContext factoryContext)
         {
+            var nullability = nullabilityContext.Create(parameter);
+            var isOptional = parameter.HasDefaultValue || nullability.ReadState == NullabilityState.Nullable;
+
             if (parameter.ParameterType == typeof(string))
             {
-                if (!parameter.HasDefaultValue)
+                factoryContext.UsingTempSourceString = true;
+
+                if (!isOptional)
                 {
-                    return valueExpression;
+                    // The following is produced if the parameter is required:
+                    //
+                    // tempSourceString = httpContext.RouteValue["param1"] ?? httpContext.Query["param1"];
+                    // if (tempSourceString == null)
+                    // {
+                    //      wasTryParseFailure = true;
+                    //      Log.RequiredParameterNotProvided(httpContext, "Int32", "param1");
+                    // }
+                    var checkRequiredStringParameterBlock = Expression.Block(
+                        Expression.Assign(TempSourceStringExpr, valueExpression),
+                        Expression.IfThen(TempSourceStringNullExpr,
+                            Expression.Block(
+                                Expression.Assign(WasTryParseFailureExpr, Expression.Constant(true)),
+                                Expression.Call(LogRequiredParameterNotProvidedMethod, 
+                                    HttpContextExpr, Expression.Constant(parameter.ParameterType.Name), Expression.Constant(parameter.Name))
+                            )
+                        )
+                    );
+
+                    factoryContext.TryParseParams.Add((TempSourceStringExpr, checkRequiredStringParameterBlock));
```

Comment:
This seems like an abuse of `factoryContext.TryParseParams`. This ends up creating unnecessary locals in `CreateTryParseCheckingResponseWritingMethodCall`. We should just add a `CheckBindingFailures` flag to FactoryContext.